src/credentialstore/credentialstore.ts

```diff
@@ -14,10 +17,25 @@ import SqlToolsServerClient from '../languageservice/serviceclient';
  */
 export class CredentialStore implements ICredentialStore {
 
-    constructor(private _client?: SqlToolsServerClient) {
+    private _useNativeCredentials: boolean;
+    private _passwordsMigrated: boolean;
+    private _secretStorage: vscode.SecretStorage;
+
+    constructor(
+        private _context: vscode.ExtensionContext,
+        private _client?: SqlToolsServerClient) {
         if (!this._client) {
             this._client = SqlToolsServerClient.instance;
         }
+        this._secretStorage = this._context.secrets;
+        this._passwordsMigrated = this._context.globalState.get(Utils.configPasswordsMigrated);
+        this._useNativeCredentials = Utils.useNativeCredentials();
+        if (this._useNativeCredentials && !this._passwordsMigrated) {
+            this.migratePasswords().then(() => {
```

Comment:
If this takes a while then that means that users may see connections fail because it tries to them up in the native store but they haven't been inserted yet. I'd suggest having a deferred that the read credentials can await on to ensure that the passwords are all migrated.